### PR TITLE
Split binary-args in lib/run.js

### PR DESCRIPTION
**Pull request by asamuzaK**
Fix [Several arguments in --binary-args doesn't work when used in jpm run command. · Issue #10](https://github.com/mozilla-jetpack/node-fx-runner/issues/10)

Split `binary-args` string into an array and concats to the `args` array.
Spaces are left as they are if the argument is quoted.
e.g. `"-a b -c \"d e\""` will concat `["-a", "b", "-c", "\"d e\""]`

### CODE REVIEW

File: lib/run.js [hunk 1]
```diff
@@ -82,3 +82,19 @@ function isProfileName (profile) {
...
+
+/**
+ * Concat binary arguments to arguments array.
+ *
+ * @param {Array} args - arguments
+ * @param {string|Array} binArgs - binary arguments
+ * @return {Array} - arguments
+ */
+function concatBinArgs (args, binArgs) {
+  args = Array.isArray(args) && args || [];
```
Comment (kumar303) [inline lib/run.js]: it would be better to use a ternary here rather than rely on assignment-by-side-effect.

File: test/run/test.run.js [hunk 2]
```diff
@@ -115,3 +116,23 @@ describe("fx-runner start", function () {
     });
   });
 });
+
+describe("concat binary arguments", function () {
+  it("binary arguments as string", function (done) {
```
Comment (kumar303) [inline test/run/test.run.js]: if you remove the `done` argument then you won't have to call done at the end of the test. By leaving out the argument, the test will magically become synchronous.

Comment (kumar303) [inline test/run/test.run.js]: I find test names easier to read when they follow the mocha convention of reading like a sentence starting with *it*. In this case it could become `it("concats binary arguments from a string", ...)`

File: lib/run.js [hunk 3]
```diff
@@ -82,3 +82,19 @@ function isProfileName (profile) {
...
+  if (typeof binArgs === "string" || binArgs instanceof String) {
+    binArgs = binArgs.match(/\"(?:[^"]|\\\")*\"|(?:[^"\s]|\")+/g);
```
Comment (kumar303) [inline lib/run.js]: I guess we should probably support single quotes too, since most shells support that?

Comment (kumar303) [inline lib/run.js]: I wonder if this should use a parser library instead, something like https://github.com/substack/node-shell-quote

Comment (kumar303) [inline lib/run.js]: bah, I dunno, that has a ton of features that we don't need

File: test/run/test.run.js [hunk 4]
```diff
@@ -115,3 +116,23 @@ describe("fx-runner start", function () {
...
+    var arr = concatBinArgs([], "-a b -c \"d e\"");
+    expect(arr[0]).to.be.equal("-a");
+    expect(arr[1]).to.be.equal("b");
+    expect(arr[2]).to.be.equal("-c");
+    expect(arr[3]).to.be.equal("\"d e\"");
```
Comment (kumar303) [inline test/run/test.run.js]: Shouldn't this equal a string of  `d e` without escaped quotes? I'm looking at the [args](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) array passed to `spawn()` and I'm pretty sure that we don't want to feed it quotation marks. Did you test this against a real binary?

**Comment by kumar303** [comment]
Hi, thanks for the patch. I know the original code was very naive but I think it's a good time to make it more robust. I added some feedback.

**Comment by asamuzaK** [comment]
Thanks for your feedback.
Updated patch.

> I wonder if this should use a parser library instead, something like https://github.com/substack/node-shell-quote

If you think it's better to use it, I'll rewrite again.

**Comment by asamuzaK** [comment]
Note: remQuotes() can also be like this, if it matches the code style.

```
  var remQuotes = function (arg) {
    var quotedStr = /^(?:"(.*)"|'(.*)')$/.exec(arg);
    return quotedStr ? (quotedStr[1] || quotedStr[2]) : arg;
  };
```

**Comment by asamuzaK** [comment]
Patch updated (v5).
* Narrowed down the matching condition of the quoted argument.
* Changed variable names.

File: lib/run.js [hunk 5]
```diff
@@ -82,3 +82,23 @@ function isProfileName (profile) {
...
+ * @return {Array} - arguments
+ */
+function concatBinArgs (args, binArgs) {
+  var removeQuotes = function (arg) {
+    var quotedArg = /^\s*(?:"(.*)"|'(.*)')\s*$/.exec(arg);
+    return quotedArg ? (quotedArg[1] || quotedArg[2]) : arg;
+  };
+  args = Array.isArray(args) ? args : [];
+  if (typeof binArgs === "string" || binArgs instanceof String) {
+    binArgs = binArgs.match(/(?:^|\s)(?:\"(?:[^"]|\\\")*\"|\'(?:[^']|\\\')*\')(?=\s|$)|(?:[^"'\s]|\"|\')+/g);
```
Comment (kumar303) [inline lib/run.js]: could you add a comment about what this is doing? Is it to handle escaped and double escaped quotes?

**Comment by kumar303** [comment]
The new regex for this is hurting my head. I think, at this point, we need to switch to a library, especially since a lot of unit test cases are missing from your patch. Did you try out [node-shell-quote](https://github.com/substack/node-shell-quote)? I think this is exactly what we need:

````javascript
var parse = require('shell-quote').parse;
var xs = parse('a "b c" \\$def \'it\\\'s great\'');
console.dir(xs);
[ 'a', 'b c', '\\$def', 'it\'s great' ]
````

It looks like their test suite [covers](https://github.com/substack/node-shell-quote/blob/master/test/parse.js) all the possibilities well.

**Comment by asamuzaK** [comment]
Updated the patch.

**Comment by kumar303** [comment]
Thanks, I think that will work a lot better. Could you also put back one of your tests that asserted a string value gets split into an array of arguments? It doesn't have to exhaustively test all edge cases -- I think we just need one test to cover the execution of the `parse()` function.

**Comment by asamuzaK** [comment]
Added test.

**Comment by kumar303** [comment]
Looks great, thanks for all the follow-ups.

**Comment by kumar303** [comment]
I just released 1.0.5 for this change https://www.npmjs.com/package/fx-runner
